File: agent/scheduler/service.py
from __future__ import annotations

import asyncio
from collections.abc import Awaitable, Callable
from datetime import datetime
from typing import Any, Literal

from loguru import logger

from ..runtime import events as runtime_events
from .models import (
    SchedulerJob,
    SchedulerPayload,
    SchedulerSchedule,
    SchedulerStatus,
    now_ms,
)
from .store import SchedulerStore, SchedulerStoreCorrupt, SchedulerStoreData
from .system_jobs import default_system_jobs


def compute_next_run_ms(schedule: SchedulerSchedule, current_ms: int) -> int | None:
    if schedule.kind == "at":
        return schedule.at_ms if schedule.at_ms and schedule.at_ms > current_ms else None
    if schedule.kind == "every":
        if not schedule.every_ms or schedule.every_ms <= 0:
            return None
        return current_ms + schedule.every_ms
    if schedule.kind == "cron":
        if not schedule.expr:
            return None
        from zoneinfo import ZoneInfo

        from croniter import croniter

        tz = ZoneInfo(schedule.tz) if schedule.tz else datetime.now().astimezone().tzinfo
        base = datetime.fromtimestamp(current_ms / 1000, tz=tz)
        next_dt = croniter(schedule.expr, base).get_next(datetime)
        return int(next_dt.timestamp() * 1000)
    return None


def validate_schedule(schedule: SchedulerSchedule) -> None:
    if schedule.kind == "at":
        if not schedule.at_ms or schedule.at_ms <= 0:
            raise ValueError("at schedule requires at_ms")
        if schedule.tz:
            raise ValueError("tz can only be used with cron schedules")
        return
    if schedule.kind == "every":
        if not schedule.every_ms or schedule.every_ms <= 0:
            raise ValueError("every schedule requires every_ms > 0")
        if schedule.tz:
            raise ValueError("tz can only be used with cron schedules")
        return
    if schedule.kind == "cron":
        if not schedule.expr:
            raise ValueError("cron schedule requires expr")
        from zoneinfo import ZoneInfo

        from croniter import croniter

        if schedule.tz:
            try:
                ZoneInfo(schedule.tz)
            except Exception:
                raise ValueError(f"unknown timezone '{schedule.tz}'") from None
        if not croniter.is_valid(schedule.expr):
            raise ValueError(f"invalid cron expression '{schedule.expr}'")
        return
    raise ValueError(f"unsupported schedule kind: {schedule.kind}")


class SchedulerService:
    def __init__(
        self,
        store: SchedulerStore,
        *,
        on_job: Callable[[SchedulerJob], Awaitable[str | None]] | None = None,
        event_sink: Callable[[dict[str, Any]], Awaitable[None]] | None = None,
        time_func: Callable[[], int] = now_ms,
        max_sleep_ms: int = 300_000,
    ):
        self.store = store
        self.on_job = on_job
        self.event_sink = event_sink
        self.time_func = time_func
        self.max_sleep_ms = max(1, int(max_sleep_ms))
        self._running = False
        self._timer_task: asyncio.Task | None = None
        self._timer_active = False
        self._loop: asyncio.AbstractEventLoop | None = None

    async def _emit(self, event: dict[str, Any]) -> None:
        if self.event_sink:
            await self.event_sink(event)

    async def start(self) -> None:
        if self._running:
            return
        self._loop = asyncio.get_running_loop()
        data = self.store.load(allow_last_good=False)
        self._register_system_jobs(data)
        self._mark_stale_running(data)
        self._running = True
        self._recompute_next_runs(data)
        self.store.save(data)
        self._arm_timer()
        logger.info("Scheduler service started with {} jobs", len(data.jobs))

    def stop(self) -> None:
        self._running = False
        task = self._timer_task
        self._timer_task = None
        if task and not task.done():
            loop = self._loop
            try:
                running_loop = asyncio.get_running_loop()
            except RuntimeError:
                running_loop = None
            if loop and not loop.is_closed() and running_loop is not loop:
                loop.call_soon_threadsafe(task.cancel)
            else:
                task.cancel()
        self._loop = None

    def list_jobs(self, *, include_disabled: bool = True) -> list[SchedulerJob]:
        return self.store.list_jobs(include_disabled=include_disabled)

    def get_job(self, job_id: str) -> SchedulerJob | None:
        return self.store.get_job(job_id)

    def status(self) -> dict[str, Any]:
        jobs = self.store.list_jobs(include_disabled=True)
        enabled = [job for job in jobs if job.enabled]
        errors = [job for job in jobs if job.state.last_status == SchedulerStatus.ERROR.value]
        return {
            "running": self._running,
            "jobs": len(jobs),
            "enabled": len(enabled),
            "nextRunAtMs": self._next_wake_ms(jobs),
            "lastError": errors[-1].state.last_error if errors else None,
        }

    def add_job(
        self,
        *,
        name: str,
        schedule: SchedulerSchedule,
        payload: SchedulerPayload,
        delete_after_run: bool = False,
        protected: bool = False,
        purpose: str | None = None,
    ) -> SchedulerJob:
        validate_schedule(schedule)
        current = self.time_func()
        job = SchedulerJob.create(
            name=name,
            schedule=schedule,
            payload=payload,
            delete_after_run=delete_after_run,
            protected=protected,
            purpose=purpose,
            now=current,
        )
        job.state.next_run_at_ms = compute_next_run_ms(schedule, current)
        self.store.upsert_job(job)
        self._arm_timer()
        return job

    def update_job(
        self,
        job_id: str,
        *,
        name: str | None = None,
        schedule: SchedulerSchedule | None = None,
        payload: SchedulerPayload | None = None,
        delete_after_run: bool | None = None,
    ) -> SchedulerJob | Literal["not_found", "protected"]:
        job = self.store.get_job(job_id)
        if job is None:
            return "not_found"
        if job.protected:
            return "protected"
        if schedule is not None:
            validate_schedule(schedule)
            job.schedule = schedule
            job.state.next_run_at_ms = compute_next_run_ms(schedule, self.time_func())
        if payload is not None:
            job.payload = payload
        if name is not None:
            job.name = str(name or "").strip() or job.name
        if delete_after_run is not None:
            job.delete_after_run = bool(delete_after_run)
        job.updated_at_ms = self.time_func()
        self.store.upsert_job(job)
        self._arm_timer()
        return job

    def enable_job(self, job_id: str, enabled: bool = True) -> SchedulerJob | Literal["not_found"]:
        job = self.store.get_job(job_id)
        if job is None:
            return "not_found"
        job.enabled = bool(enabled)
        job.updated_at_ms = self.time_func()
        job.state.next_run_at_ms = (
            compute_next_run_ms(job.schedule, self.time_func()) if job.enabled else None
        )
        self.store.upsert_job(job)
        self._arm_timer()
        return job

    def remove_job(self, job_id: str) -> SchedulerJob | Literal["not_found", "protected"]:
        job = self.store.get_job(job_id)
        if job is None:
            return "not_found"
        if job.protected:
            return "protected"
        removed = self.store.remove_job(job_id)
        self._arm_timer()
        return removed or "not_found"

    async def run_job(self, job_id: str, *, force: bool = False) -> bool:
        job = self.store.get_job(job_id)
        if job is None:
            return False
        if not force and not job.enabled:
            return False
        data = self.store.load()
        live = next((item for item in data.jobs if item.id == job.id), None)
        if live is None:
            return False
        await self._execute_job(live, data, manual=True)
        self.store.save(data)
        if self._running:
            self._arm_timer()
        return True

    def _recompute_next_runs(self, data: SchedulerStoreData) -> None:
        current = self.time_func()
        for job in data.jobs:
            if job.enabled:
                job.state.next_run_at_ms = compute_next_run_ms(job.schedule, current)
            else:
                job.state.next_run_at_ms = None

    def _register_system_jobs(self, data: SchedulerStoreData) -> None:
        current = self.time_func()
        existing = {job.id: job for job in data.jobs}
        for default in default_system_jobs(now=current):
            found = existing.get(default.id)
            if found is None:
                default.state.next_run_at_ms = compute_next_run_ms(default.schedule, current)
                data.jobs.append(default)
                continue
            found.name = default.name
            found.schedule = default.schedule
            found.payload = default.payload
            found.protected = True
            found.purpose = default.purpose
            found.delete_after_run = False
            found.updated_at_ms = current
            if found.enabled:
                found.state.next_run_at_ms = compute_next_run_ms(found.schedule, current)

    def _next_wake_ms(self, jobs: list[SchedulerJob] | None = None) -> int | None:
        jobs = jobs if jobs is not None else self.store.list_jobs(include_disabled=True)
        times = [
            job.state.next_run_at_ms for job in jobs
            if job.enabled and job.state.next_run_at_ms
        ]
        return min(times) if times else None

    def _arm_timer(self) -> None:
        if not self._running:
            return
        try:
            running_loop = asyncio.get_running_loop()
        except RuntimeError:
            running_loop = None
        target_loop = running_loop or self._loop
        if target_loop is None or target_loop.is_closed():
            logger.warning("Scheduler timer could not be armed: no active event loop")
            return
        if running_loop is target_loop:
            self._arm_timer_on_loop()
            return
        target_loop.call_soon_threadsafe(self._arm_timer_on_loop)

    def _arm_timer_on_loop(self) -> None:
        if not self._running:
            return
        if self._timer_task:
            self._timer_task.cancel()
        next_wake = self._next_wake_ms()
        delay_ms = self.max_sleep_ms if next_wake is None else min(
            self.max_sleep_ms,
            max(0, next_wake - self.time_func()),
        )

        async def tick() -> None:
            try:
                await asyncio.sleep(delay_ms / 1000)
                if self._running:
                    await self._on_timer()
            except asyncio.CancelledError:
                pass

        self._timer_task = asyncio.create_task(tick())

    async def _on_timer(self) -> None:
        try:
            data = self.store.load()
        except SchedulerStoreCorrupt:
            logger.warning("Scheduler store corrupt during timer tick; keeping service alive")
            self._arm_timer()
            return
        self._timer_active = True
        try:
            current = self.time_func()
            due = [
                job for job in data.jobs
                if job.enabled and job.state.next_run_at_ms and current >= job.state.next_run_at_ms
            ]
            for job in due:
                await self._execute_job(job, data)
        finally:
            self._timer_active = False
        self._arm_timer()

    async def _execute_job(
        self,
        job: SchedulerJob,
        data: SchedulerStoreData,
        *,
        manual: bool = False,
    ) -> None:
        start = self.time_func()
        error: str | None = None
        status = SchedulerStatus.OK.value
        job.state.last_run_at_ms = start
        job.state.last_status = SchedulerStatus.RUNNING.value
        job.state.last_error = None
        job.state.next_run_at_ms = None
        job.updated_at_ms = start
        self.store.save(data)
        await self._emit(runtime_events.scheduler_run_start(job.to_dict()))
        try:
            if self.on_job:
                await self.on_job(job)
        except asyncio.CancelledError:
            status = SchedulerStatus.CANCELLED.value
            error = "cancelled"
            logger.info("Scheduler: job '{}' cancelled", job.name)
        except Exception as exc:
            status = SchedulerStatus.ERROR.value
            error = str(exc)
            logger.error("Scheduler: job '{}' failed: {}", job.name, exc)

        end = self.time_func()
        job.state.record_run(
            run_at_ms=start,
            status=status,
            duration_ms=max(0, end - start),
            error=error,
        )
        job.updated_at_ms = end

        if job.schedule.kind == "at" and not manual:
            if job.delete_after_run:
                data.jobs = [item for item in data.jobs if item.id != job.id]
            else:
                job.enabled = False
                job.state.next_run_at_ms = None
        elif job.enabled:
            job.state.next_run_at_ms = compute_next_run_ms(job.schedule, self.time_func())
        else:
            job.state.next_run_at_ms = None
        if status == SchedulerStatus.ERROR.value:
            await self._emit(
                runtime_events.scheduler_run_error(job.to_dict(), error=error or "unknown error")
            )
        elif status == SchedulerStatus.CANCELLED.value:
            await self._emit(
                runtime_events.scheduler_run_cancelled(job.to_dict(), reason=error or "cancelled")
            )
        else:
            await self._emit(runtime_events.scheduler_run_done(job.to_dict()))
        self.store.save(data)

    def _mark_stale_running(self, data: SchedulerStoreData) -> None:
        current = self.time_func()
        for job in data.jobs:
            if job.state.last_status != SchedulerStatus.RUNNING.value:
                continue
            started = job.state.last_run_at_ms or current
            error = "interrupted by scheduler restart"
            job.state.record_run(
                run_at_ms=started,
                status=SchedulerStatus.ERROR.value,
                duration_ms=max(0, current - started),
                error=error,
            )
            job.updated_at_ms = current

File: packages/core/src/scheduler/service.ts
import * as runtimeEvents from '../runtime/events'
import { cleanString } from '../util/strings'
import {
  computeNextRunMs,
  SchedulerJob,
  SchedulerPayload,
  SchedulerSchedule,
  SchedulerStatus,
  SCHEDULER_TARGET_SESSION_METADATA_KEY,
  nowMs,
  schedulerPayloadSessionId,
  validateSchedule,
} from './models'
import {
  SchedulerStore,
  SchedulerStoreCorrupt,
  SchedulerStoreData,
} from './store'
import { defaultSystemJobs } from './system-jobs'

export type SchedulerTimerCallback = () => void | Promise<void>
export type SchedulerSetTimer = (
  callback: SchedulerTimerCallback,
  delayMs: number,
) => unknown
export type SchedulerClearTimer = (handle: unknown) => void
export type SchedulerTargetSession = () => string | null | undefined

export class SchedulerService {
  readonly store: SchedulerStore
  onJob: ((job: SchedulerJob) => Promise<string | void | null>) | null
  eventSink: ((event: Record<string, unknown>) => Promise<void>) | null
  timeFunc: () => number
  maxSleepMs: number
  private readonly targetSessionId: SchedulerTargetSession
  private readonly setTimer: SchedulerSetTimer
  private readonly clearTimer: SchedulerClearTimer
  private running = false
  private timer: unknown = null

  constructor(
    store: SchedulerStore,
    opts: {
      onJob?: ((job: SchedulerJob) => Promise<string | void | null>) | null
      eventSink?: ((event: Record<string, unknown>) => Promise<void>) | null
      timeFunc?: () => number
      maxSleepMs?: number
      targetSessionId?: SchedulerTargetSession | null
      setTimer?: SchedulerSetTimer
      clearTimer?: SchedulerClearTimer
    } = {},
  ) {
    this.store = store
    this.onJob = opts.onJob ?? null
    this.eventSink = opts.eventSink ?? null
    this.timeFunc = opts.timeFunc ?? nowMs
    this.maxSleepMs = Math.max(1, Math.trunc(opts.maxSleepMs ?? 300_000))
    this.targetSessionId = opts.targetSessionId ?? (() => null)
    this.setTimer =
      opts.setTimer ??
      ((callback, delayMs) =>
        setTimeout(() => {
          void callback()
        }, delayMs))
    this.clearTimer =
      opts.clearTimer ??
      ((handle) => {
        clearTimeout(handle as ReturnType<typeof setTimeout>)
      })
  }

  async start(): Promise<void> {
    if (this.running) return
    const data = this.store.load({ allowLastGood: false })
    this.registerSystemJobs(data)
    this.markStaleRunning(data)
    this.running = true
    this.recomputeNextRuns(data)
    this.store.save(data)
    this.armTimer(data)
  }

  stop(): void {
    this.running = false
    if (this.timer !== null) {
      this.clearTimer(this.timer)
      this.timer = null
    }
  }

  listJobs(opts: { includeDisabled?: boolean } = {}): SchedulerJob[] {
    return this.store.listJobs({
      includeDisabled: opts.includeDisabled ?? true,
    })
  }
  getJob(jobId: string): SchedulerJob | null {
    return this.store.getJob(jobId)
  }
  status(): Record<string, unknown> {
    const jobs = this.store.listJobs({ includeDisabled: true })
    const enabled = jobs.filter((job) => job.enabled)
    const errors = jobs.filter(
      (job) => job.state.last_status === SchedulerStatus.ERROR,
    )
    return {
      running: this.running,
      jobs: jobs.length,
      enabled: enabled.length,
      nextRunAtMs: this.nextWakeMs(jobs),
      lastError: errors.at(-1)?.state.last_error ?? null,
    }
  }

  addJob(opts: {
    name: string
    schedule: SchedulerSchedule
    payload: SchedulerPayload
    deleteAfterRun?: boolean
    protected?: boolean
    purpose?: string | null
  }): SchedulerJob {
    validateSchedule(opts.schedule)
    const current = this.timeFunc()
    const payload = this.withTargetSession(opts.payload)
    const job = SchedulerJob.create({
      name: opts.name,
      schedule: opts.schedule,
      payload,
      deleteAfterRun: opts.deleteAfterRun ?? false,
      protected: opts.protected ?? false,
      purpose: opts.purpose ?? null,
      now: current,
    })
    job.state.next_run_at_ms = computeNextRunMs(opts.schedule, current)
    const saved = this.store.upsertJob(job)
    this.armTimer()
    return saved
  }

  updateJob(
    jobId: string,
    opts: {
      name?: string | null
      schedule?: SchedulerSchedule | null
      payload?: SchedulerPayload | null
      deleteAfterRun?: boolean | null
    },
  ): SchedulerJob | 'not_found' | 'protected' {
    const job = this.store.getJob(jobId)
    if (!job) return 'not_found'
    if (job.protected) return 'protected'
    if (opts.schedule) {
      validateSchedule(opts.schedule)
      job.schedule = opts.schedule
      job.state.next_run_at_ms = computeNextRunMs(job.schedule, this.timeFunc())
    }
    if (opts.payload) job.payload = opts.payload
    if (opts.name !== undefined && opts.name !== null)
      job.name = String(opts.name || '').trim() || job.name
    if (opts.deleteAfterRun !== undefined && opts.deleteAfterRun !== null)
      job.delete_after_run = Boolean(opts.deleteAfterRun)
    job.updated_at_ms = this.timeFunc()
    const saved = this.store.upsertJob(job)
    this.armTimer()
    return saved
  }

  enableJob(jobId: string, enabled = true): SchedulerJob | 'not_found' {
    const job = this.store.getJob(jobId)
    if (!job) return 'not_found'
    job.enabled = Boolean(enabled)
    job.updated_at_ms = this.timeFunc()
    job.state.next_run_at_ms = job.enabled
      ? computeNextRunMs(job.schedule, this.timeFunc())
      : null
    const saved = this.store.upsertJob(job)
    this.armTimer()
    return saved
  }

  removeJob(jobId: string): SchedulerJob | 'not_found' | 'protected' {
    const job = this.store.getJob(jobId)
    if (!job) return 'not_found'
    if (job.protected) return 'protected'
    const removed = this.store.removeJob(jobId) ?? 'not_found'
    this.armTimer()
    return removed
  }

  async runJob(
    jobId: string,
    opts: { force?: boolean } = {},
  ): Promise<boolean> {
    const job = this.store.getJob(jobId)
    if (!job) return false
    if (!opts.force && !job.enabled) return false
    await this.executeJob(job, { manual: true })
    this.armTimer()
    return true
  }

  async onTimer(): Promise<void> {
    let data: SchedulerStoreData
    try {
      data = this.store.load()
    } catch (error) {
      if (error instanceof SchedulerStoreCorrupt) return
      throw error
    }
    const current = this.timeFunc()
    for (const job of data.jobs.filter(
      (item) =>
        item.enabled &&
        item.state.next_run_at_ms &&
        current >= item.state.next_run_at_ms,
    )) {
      await this.executeJob(job, { manual: false })
    }
    this.armTimer()
  }

  private armTimer(data?: SchedulerStoreData): void {
    if (!this.running) return
    if (this.timer !== null) this.clearTimer(this.timer)
    const jobs = data?.jobs ?? this.store.listJobs({ includeDisabled: true })
    const nextWake = this.nextWakeMs(jobs)
    const delayMs =
      nextWake === null
        ? this.maxSleepMs
        : Math.min(this.maxSleepMs, Math.max(0, nextWake - this.timeFunc()))
    this.timer = this.setTimer(() => this.onTimer(), delayMs)
  }

  private recomputeNextRuns(data: SchedulerStoreData): void {
    const current = this.timeFunc()
    for (const job of data.jobs)
      job.state.next_run_at_ms = job.enabled
        ? computeNextRunMs(job.schedule, current)
        : null
  }

  private registerSystemJobs(data: SchedulerStoreData): void {
    const current = this.timeFunc()
    const existing = new Map(data.jobs.map((job) => [job.id, job]))
    for (const def of defaultSystemJobs(current)) {
      const found = existing.get(def.id)
      if (!found) {
        def.state.next_run_at_ms = computeNextRunMs(def.schedule, current)
        data.jobs.push(def)
      } else {
        found.name = def.name
        found.schedule = def.schedule
        found.payload = def.payload
        found.protected = true
        found.purpose = def.purpose
        found.delete_after_run = false
        if (found.enabled)
          found.state.next_run_at_ms = computeNextRunMs(found.schedule, current)
      }
    }
  }

  private nextWakeMs(jobs: SchedulerJob[]): number | null {
    const times = jobs
      .filter((job) => job.enabled && job.state.next_run_at_ms)
      .map((job) => job.state.next_run_at_ms!)
    return times.length ? Math.min(...times) : null
  }

  /**
   * 单个 job 的执行状态通过 appendAction 追加到 action log，而不是整表 save()。
   * onTimer/runJob 的 `data` 快照可能跨越 `await this.onJob(job)` 变陈旧；若在此期间
   * 整表覆盖写，会悄悄撤销执行期间发生的并发 add/update/remove（审计 P0-3）。
   * appendAction 只记录"这一个 job 的最新状态"，下一次 store.load() 会自动合并，
   * 不会覆盖其他 job 的并发变更。
   */
  private async executeJob(
    job: SchedulerJob,
    opts: { manual: boolean },
  ): Promise<void> {
    const start = this.timeFunc()
    let status: string = SchedulerStatus.OK
    let error: string | null = null
    job.state.last_run_at_ms = start
    job.state.last_status = SchedulerStatus.RUNNING
    job.state.last_error = null
    job.state.next_run_at_ms = null
    job.updated_at_ms = start
    this.store.appendAction('update', { job })
    await this.emit(
      this.withEventSession(runtimeEvents.schedulerRunStart(job.toDict()), job),
    )
    try {
      if (this.onJob) await this.onJob(job)
    } catch (exc) {
      status =
        exc instanceof Error && exc.name === 'CancelledTaskError'
          ? SchedulerStatus.CANCELLED
          : SchedulerStatus.ERROR
      error =
        status === SchedulerStatus.CANCELLED
          ? 'cancelled'
          : String(exc instanceof Error ? exc.message : exc)
    }
    const end = this.timeFunc()
    job.state.recordRun({
      runAtMs: start,
      status,
      durationMs: Math.max(0, end - start),
      error,
    })
    job.updated_at_ms = end
    let deleted = false
    if (job.schedule.kind === 'at' && !opts.manual) {
      if (job.delete_after_run) deleted = true
      else {
        job.enabled = false
        job.state.next_run_at_ms = null
      }
    } else if (job.enabled)
      job.state.next_run_at_ms = computeNextRunMs(job.schedule, this.timeFunc())
    this.store.appendAction(
      deleted ? 'delete' : 'update',
      deleted ? { jobId: job.id } : { job },
    )
    if (status === SchedulerStatus.ERROR)
      await this.emit(
        this.withEventSession(
          runtimeEvents.schedulerRunError(job.toDict(), {
            error: error || 'unknown error',
          }),
          job,
        ),
      )
    else if (status === SchedulerStatus.CANCELLED)
      await this.emit(
        this.withEventSession(
          runtimeEvents.schedulerRunCancelled(job.toDict(), {
            reason: error || 'cancelled',
          }),
          job,
        ),
      )
    else
      await this.emit(
        this.withEventSession(
          runtimeEvents.schedulerRunDone(job.toDict()),
          job,
        ),
      )
  }

  private markStaleRunning(data: SchedulerStoreData): void {
    const current = this.timeFunc()
    for (const job of data.jobs) {
      if (job.state.last_status !== SchedulerStatus.RUNNING) continue
      const started = job.state.last_run_at_ms ?? current
      job.state.recordRun({
        runAtMs: started,
        status: SchedulerStatus.ERROR,
        durationMs: Math.max(0, current - started),
        error: 'interrupted by scheduler restart',
      })
      job.updated_at_ms = current
    }
  }

  private async emit(event: Record<string, unknown>): Promise<void> {
    if (this.eventSink) await this.eventSink(event)
  }

  private withTargetSession(payload: SchedulerPayload): SchedulerPayload {
    if (schedulerPayloadSessionId(payload)) return payload
    const sessionId = cleanString(this.targetSessionId())
    if (!sessionId) return payload
    return SchedulerPayload.fromDict({
      ...payload.toDict(),
      meta: {
        ...payload.meta,
        [SCHEDULER_TARGET_SESSION_METADATA_KEY]: sessionId,
      },
    })
  }

  /** 兜底链 payload → 实时 targetSessionId → 伪 session 'scheduler'：绝不发无主事件。 */
  private withEventSession(
    event: Record<string, unknown>,
    job: SchedulerJob,
  ): Record<string, unknown> {
    const sessionId =
      schedulerPayloadSessionId(job.payload) ||
      cleanString(this.targetSessionId()) ||
      'scheduler'
    return { ...event, session_id: sessionId }
  }
}

export {
  computeNextRunMs,
  validateSchedule,
  SchedulerJob,
  SchedulerPayload,
  SchedulerSchedule,
  SchedulerStatus,
}
